File: src/types/domain/reasoning.ts
/**
 * Reasoning/Thinking context types for multi-provider support.
 *
 * This module provides unified data structures for handling reasoning content
 * from different LLM providers (Anthropic Extended Thinking, OpenAI Reasoning).
 *
 * Design based on opencode-dev's reasoning architecture.
 */

/**
 * Unified reasoning part that works across all providers.
 */
export interface ReasoningPart {
  /** Unique identifier for this reasoning part */
  id: string
  /** Session ID this part belongs to */
  sessionId: number
  /** Message ID this part belongs to */
  messageId: string
  /** The reasoning/thinking text content */
  text: string
  /** Provider-specific metadata */
  metadata?: ReasoningMetadata
  /** Timing information */
  time: ReasoningTime
}

/**
 * Provider-specific metadata for reasoning parts.
 */
export interface ReasoningMetadata {
  /** OpenAI Responses API item ID (for item_reference) */
  itemId?: string
  /** OpenAI encrypted reasoning content */
  encryptedContent?: string
  /** Anthropic thinking signature for verification */
  signature?: string
  /** Provider identifier (e.g., "openai", "anthropic") */
  provider?: string
}

/**
 * Timing information for reasoning parts.
 */
export interface ReasoningTime {
  /** When reasoning started (Unix timestamp ms) */
  start: number
  /** When reasoning ended (Unix timestamp ms), undefined if still in progress */
  end?: number
}

/**
 * Stream event types for reasoning processing.
 */
export type ReasoningEvent = ReasoningStartEvent | ReasoningDeltaEvent | ReasoningEndEvent

export interface ReasoningStartEvent {
  type: 'reasoning_start'
  /** Unique ID for this reasoning block */
  id: string
  /** Provider-specific metadata */
  metadata?: ReasoningMetadata
}

export interface ReasoningDeltaEvent {
  type: 'reasoning_delta'
  /** ID of the reasoning block */
  id: string
  /** Text delta to append */
  delta: string
  /** Optional metadata update */
  metadata?: ReasoningMetadata
}

export interface ReasoningEndEvent {
  type: 'reasoning_end'
  /** ID of the reasoning block */
  id: string
  /** Final metadata (may include signature) */
  metadata?: ReasoningMetadata
}

/**
 * Helper functions for reasoning parts
 */
export const ReasoningUtils = {
  /**
   * Create a new reasoning part with start time.
   */
  create(id: string, sessionId: number, messageId: string): ReasoningPart {
    return {
      id,
      sessionId,
      messageId,
      text: '',
      time: {
        start: Date.now(),
      },
    }
  },

  /**
   * Check if reasoning is still in progress.
   */
  isStreaming(part: ReasoningPart): boolean {
    return part.time.end === undefined
  },

  /**
   * Get duration in milliseconds if completed.
   */
  durationMs(part: ReasoningPart): number | undefined {
    if (part.time.end === undefined) return undefined
    return part.time.end - part.time.start
  },

  /**
   * Get OpenAI item_id for item_reference.
   */
  getOpenAIItemId(part: ReasoningPart): string | undefined {
    return part.metadata?.itemId
  },

  /**
   * Get Anthropic signature.
   */
  getAnthropicSignature(part: ReasoningPart): string | undefined {
    return part.metadata?.signature
  },

  /**
   * Check if reasoning has OpenAI metadata.
   */
  isOpenAI(part: ReasoningPart): boolean {
    return part.metadata?.provider === 'openai'
  },

  /**
   * Check if reasoning has Anthropic metadata.
   */
  isAnthropic(part: ReasoningPart): boolean {
    return part.metadata?.provider === 'anthropic'
  },
}

File: src-tauri/src/agent/types/reasoning.rs
//! Reasoning/Thinking context types for multi-provider support.
//!
//! Based on opencode-dev's design: unified ReasoningPart with provider-specific metadata.

use chrono::Utc;
use serde::{Deserialize, Serialize};

// Re-export the single metadata type to avoid duplication
pub use crate::llm::anthropic_types::ReasoningBlockMetadata as ReasoningMetadata;

/// Unified reasoning part that works across all providers.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReasoningPart {
    pub id: String,
    pub session_id: i64,
    pub message_id: String,
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<ReasoningMetadata>,
    pub time: ReasoningTime,
}

/// Timing information for reasoning parts.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReasoningTime {
    pub start: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub end: Option<i64>,
}

impl ReasoningPart {
    pub fn new(id: String, session_id: i64, message_id: String) -> Self {
        Self {
            id,
            session_id,
            message_id,
            text: String::new(),
            metadata: None,
            time: ReasoningTime {
                start: Utc::now().timestamp_millis(),
                end: None,
            },
        }
    }

    pub fn append_text(&mut self, delta: &str) {
        self.text.push_str(delta);
    }

    pub fn complete(&mut self) {
        self.time.end = Some(Utc::now().timestamp_millis());
        // Trim in place to avoid allocation
        let trimmed_len = self.text.trim_end().len();
        self.text.truncate(trimmed_len);
    }

    pub fn is_streaming(&self) -> bool {
        self.time.end.is_none()
    }

    pub fn set_openai_metadata(&mut self, item_id: String, encrypted_content: Option<String>) {
        let m = self.metadata.get_or_insert_with(ReasoningMetadata::default);
        m.item_id = Some(item_id);
        m.encrypted_content = encrypted_content;
        m.provider = Some("openai".to_string());
    }

    pub fn set_anthropic_metadata(&mut self, signature: Option<String>) {
        let m = self.metadata.get_or_insert_with(ReasoningMetadata::default);
        m.signature = signature;
        m.provider = Some("anthropic".to_string());
    }

    pub fn openai_item_id(&self) -> Option<&str> {
        self.metadata.as_ref().and_then(|m| m.item_id.as_deref())
    }

    pub fn anthropic_signature(&self) -> Option<&str> {
        self.metadata.as_ref().and_then(|m| m.signature.as_deref())
    }
}
